services/housekeeper/housekeeper.go

```diff
@@ -196,27 +207,35 @@ func (hk *Housekeeper) updateKnownValidators() {
 	printCounter := len(hk.proposersAlreadySaved) == 0 // only on first round
 	i := 0
 	newValidators := 0
+	bufferSize := 10000
+	indexPkMap := make(map[uint64]types.PubkeyHex)
 	for _, validator := range validators {
 		i++
-		if printCounter && i%10000 == 0 {
-			hk.log.Debugf("writing to redis: %d / %d", i, numValidators)
-		}
 
 		// avoid resaving if index->pubkey mapping is the same
 		prevPubkeyForIndex := hk.proposersAlreadySaved[validator.Index]
 		if prevPubkeyForIndex == validator.Validator.Pubkey {
 			continue
 		}
 
-		err := hk.redis.SetKnownValidator(types.PubkeyHex(validator.Validator.Pubkey), validator.Index)
-		if err != nil {
-			log.WithError(err).WithField("pubkey", validator.Validator.Pubkey).Error("failed to set known validator in Redis")
-		} else {
-			hk.proposersAlreadySaved[validator.Index] = validator.Validator.Pubkey
-			newValidators++
+		indexPkMap[validator.Index] = types.PubkeyHex(validator.Validator.Pubkey)
+
+		if i%bufferSize == 0 {
+			hk.flushKnownValidators(indexPkMap)
+			indexPkMap = make(map[uint64]types.PubkeyHex)
+			newValidators += bufferSize
+			if printCounter {
+				hk.log.Debugf("writing to redis: %d / %d", i, numValidators)
+			}
 		}
 	}
 
+	hk.flushKnownValidators(indexPkMap)
+	newValidators += len(indexPkMap)
+	if printCounter {
```

Comment:
i think that loop can just be removed, before it was more of a progress update log, because it took like a minute

Reply:
I know :-)
Although Redis can handle 500K records in a single call, uploading chunks in the name of resiliency is better.
This way, it's not all or nothing. You might lose some but not all of it.